plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -590,25 +595,118 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
```

Comment:
Ok, so here's the high level expected control flow that I _think_ correctly captures the contract implementation and is more straightforward to reason about (although I think more cleanup is possible):
```
ObjectMetadata metadata = getS3ObjectMetadata(path);
if (recursive) {
  int deletedCount = 0;
  if (metadata == null || path.getName().endsWith("_$folder$") || MediaType.parse(metadata.getContentType()).is(DIRECTORY_MEDIA_TYPE)) {
    // directory (including "virtual" when metadata == null)
    deletedCount += deletePrefix(path + PATH_SEPARATOR); // only delete keys with: "<prefix>/" not "<prefix>_sibbling"
  }
  if (metadata != null && deleteObject(keyFromPath(path)) {
    deletedCount++; // "directory" object successfully deleted
  }
  return deletedCount > 0; // must delete at least one object to count as successful
}
else {
  if (metadata == null) {
    if (path.getName().isEmpty()) {
      return false; // bucket root does not exist
    }
    if (listPath(path, OptionalInt.of(1), SHALLOW_ALL).hasNext()) {
      return false; // non-empty directory
    }
    return deleteObject(keyFromPath(path) + PATH_SEPARATOR); // delete empty directory entry if present
  }
  else if (path.getName().endsWith("_$folder$") || MediaType.parse(metadata.getContentType()).is(DIRECTORY_MEDIA_TYPE)) {
    if (listPath(path, OptionalInt.of(1), SHALLOW_ALL).hasNext()) {
      return false; // non-empty directory
    }
  }
  return deleteObject(keyFromPath(path)); // file or empty directory deletion
}
```

Reply:
Thank you @pettyjamesm for coming up with this algorithm.

It makes indeed things a bit more clear when doing deletes.

I took the time to implement the algorithm and also add corresponding tests for AWS and also MinIO to make sure the functionality is provided.